lib/web_ui/lib/src/engine/text_editing/text_editing.dart

```diff
@@ -440,6 +440,202 @@ class AutofillInfo {
   }
 }
 
+/// Replaces a range of text in the original string with the text given in the
+/// replacement string.
+String _replace(String originalText, String replacementText, ui.TextRange replacedRange) {
+  assert(replacedRange.isValid);
+  assert(replacedRange.start <= originalText.length && replacedRange.end <= originalText.length);
+
+  final ui.TextRange normalizedRange = ui.TextRange(start: math.min(replacedRange.start, replacedRange.end), end: math.max(replacedRange.start, replacedRange.end));
+
+  final String newText = normalizedRange.textBefore(originalText) + replacementText + normalizedRange.textAfter(originalText);
+
+  return newText;
+}
+
+/// The change between the last editing state and the current editing state
+/// of a text field. This is packaged into a JSON and sent to the framework
+/// to be processed into a concrete [TextEditingDelta].
+class TextEditingDeltaState {
+  TextEditingDeltaState({
+    this.oldText = '',
+    this.deltaText = '',
+    this.deltaStart = -1,
+    this.deltaEnd = -1,
+    this.baseOffset = -1,
+    this.extentOffset = -1,
+    this.composingOffset = -1,
+    this.composingExtent = -1,
+  });
+
+  /// Infers the correct delta values based on information from the new editing state
+  /// and the last editing state.
+  ///
+  /// For a deletion we calculate the length of the deleted text by comparing the new
+  /// and last editing states. We subtract this from the deltaEnd that we set when beforeinput
+  /// was fired to determine out deltaStart.
+  ///
+  /// For a replacement at a selection we set the deltaStart to be the beginning of the selection
+  /// from the last editing state.
+  ///
+  /// For the composing region we check if a composing range was captured by the compositionupdate event,
+  /// we have a non empty deltaText, and that we did not have an active selection. An active selection
+  /// would mean we are not composing.
+  ///
+  /// We then verify that the delta we collected results in the text contained within the new editing state
+  /// when applied to the last editing state. If it is not then we use our new editing state as the source of truth,
+  /// and use regex to find the correct deltaStart and deltaEnd.
+  static TextEditingDeltaState inferDeltaState(EditingState newEditingState, EditingState? lastEditingState, TextEditingDeltaState lastTextEditingDeltaState) {
+    final TextEditingDeltaState newTextEditingDeltaState = lastTextEditingDeltaState.copyWith();
```

Comment:
Thanks for doing this, I think it's comforting to know there's nothing fancy going on with modifying the inputs when I see this copy.